formatter/src/main/java/org/togetherjava/formatter/Formatter.java

```diff
@@ -0,0 +1,150 @@
+package org.togetherjava.formatter;
+
+import org.togetherjava.formatter.tokenizer.Lexer;
+import org.togetherjava.formatter.tokenizer.Token;
+import org.togetherjava.formatter.tokenizer.TokenType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Formatter which can format a given string into a string which contains code blocks etc
+ *
+ * @author illuminator3
+ */
+public class Formatter {
+    /**
+     * Formats the given tokens
+     *
+     * @param tokens tokens to format
+     * @return resulting code
+     * @author illuminator3
+     */
+    public String format(List<Token> tokens) {
+        List<Section> sections = sectionize(indexTokens(tokens));
+        StringBuilder result = new StringBuilder();
+
+        for (Section section : sections) {
+            if (section.isCodeSection()) {
+                result.append("```java\n")
+                    .append(writeCodeSection(section.tokens()))
+                    .append("\n```");
+            } else {
+                result.append(joinTokens(section.tokens()));
+            }
+        }
+
+        return result.toString();
+    }
+
+    /**
+     * Formats the given string using a given lexer
+     *
+     * @param input input to format
+     * @param lexer lexer to use
+     * @return resulting code
+     * @author illuminator3
+     */
+    public String format(String input, Lexer lexer) {
+        return format(lexer.tokenize(input));
+    }
+
+    /**
+     * Joins given tokens together and normalizes whitespaces
+     *
+     * @param tokens tokens to join
+     * @return joined form of the tokens
+     * @author illuminator3
+     */
+    private String joinTokens(List<Token> tokens) {
+        return tokens.stream().map(Token::content).collect(Collectors.joining());
+    }
+
+    /**
+     * Writes and formats a given code section (in form of a list of tokens) into a StringBuilder
+     * using a {@link CodeSectionFormatter}
+     *
+     * @param tokens tokens to write
+     * @return written code sections
+     * @author illuminator3
+     */
+    private StringBuilder writeCodeSection(List<Token> tokens) {
+        CodeSectionFormatter formatter = new CodeSectionFormatter(tokens);
+
+        formatter.format();
+
+        return formatter.result();
```

Comment:
Why new lines after each line in here, it's all the same context, it should be in the same section.
```suggestion
    private StringBuilder writeCodeSection(List<Token> tokens) {
        CodeSectionFormatter formatter = new CodeSectionFormatter(tokens);
        formatter.format();
        return formatter.result();
```

Reply:
subjective.